### PR TITLE
Update toposort.jl

### PR DESCRIPTION
**Pull request by millerjoey**
Tests are failing since SimpleGraphs 0.5.2. Changing to see if tests pass now. Will we need to also update compat?

### CODE REVIEW

**Comment by cscherrer** [comment]
Good catch, thanks @millerjoey ! 

You're talking about depending on an older `SimpleGraphs`, right? Yes, we'd need to update `compat` for 0.13. Hopefully we could get 0.14 to depend on the latest `SimpleGraphs`

**Comment by millerjoey** [comment]
Yeah. Eltype doesn't exist on any previous versions of SimpleGraphs, so I
think we need to remove the 0.3 and 0.4 entries. Do we manually do that?

**Comment by cscherrer** [comment]
Yeah, I think this would be manual. Big question to me is... would this go into 0.13 or 0.14? We should get @DilumAluthge 's take on this too

**Comment by millerjoey** [comment]
The error that causes this to fail (CI and on my laptop running `using Soss` on Julia 1.5) is strange to me. It looks like somewhere the order of `includes` or struct definitions isn't correct anymore. I don't understand why this PR causes that to fail, since I have no problem `using Soss` on master right now. Any thoughts @DilumAluthge or @cscherrer ?

**Comment by cscherrer** [comment]
I think it may be because `statements.jl` is `include`d before `model.jl`, but the end of `statement.jl` has

```julia
function statements(m::Model)
    Statement[Soss.findStatement(m, v) for v in variables(m)]
end
```

Guess we might need to move that to the end of `model.jl`

**Comment by cscherrer** [comment]
Thanks @millerjoey !

**Comment by cscherrer** [comment]
Looks like now it's failing a doctest. @DilumAluthge had pointed out that PRNG consistency may not be guaranteed between Julia versions, that's probably what's happening here. Maybe we could use the same seed _and_ manually specify the RNG to use?

**Comment by DilumAluthge** [comment]
That won't be sufficient unfortunately. The only way to have stable RNG output across different minor Julia versions is to use the StableRNGs.jl package.

**Comment by DilumAluthge** [comment]
> manually specify the RNG to use?

Ah I just saw this part. Yes we have to manually specify the RNG, but it needs to be from the StableRNGs.jl package.

**Comment by cscherrer** [comment]
Right, I just saw [this](https://docs.julialang.org/en/v1/stdlib/Random/#Reproducibility) too:
> By using an RNG parameter initialized with a given seed, you can reproduce the same pseudorandom number sequence when running your program multiple times. However, a minor release of Julia (e.g. 1.3 to 1.4) may change the sequence of pseudorandom numbers generated from a specific seed. (Even if the sequence produced by a low-level function like rand does not change, the output of higher-level functions like randsubseq may change due to algorithm updates.) Rationale: guaranteeing that pseudorandom streams never change prohibits many algorithmic improvements.

**Comment by cscherrer** [comment]
Maybe we should try StableRNGs as a test dependency, hopefully that might do it

**Comment by DilumAluthge** [comment]
I think you have to construct a `StableRNG` and then explicitly pass it to every function that uses an RNG.

**Comment by millerjoey** [comment]
Just ahead of you guys. Already added `StableRNGs` and modified the dynamichmc test ;)

Unfortunately advancedhmc doesn't let me pass the rng so I had to undoctest that one.

Now I'm just fixing an error on the markdown test, since `weave` has changed since it last worked.

**Comment by DilumAluthge** [comment]
> Unfortunately advancedhmc doesn't let me pass the rng so I had to undoctest that one.

@millerjoey Could you open an issue on AdvancedHMC.jl to discuss this? I think all sampling packages should allow the user to specify the RNG, like DynamicHMC.jl does.

**Comment by cscherrer** [comment]
I guess we'll need a `rand(RNG, ::JointDistribution)` too. Probably best to have everything there with RNG first argument, then have a default method if no RNG is passed

**Comment by DilumAluthge** [comment]
> I guess we'll need a `rand(RNG, ::JointDistribution)` too. Probably best to have everything there with RNG first argument, then have a default method if no RNG is passed

Yep exactly.

**Comment by DilumAluthge** [comment]
We'll also need to add those methods in SossMLJ.jl, just for completeness sake.

**Comment by millerjoey** [comment]
> Could you open an issue on AdvancedHMC.jl to discuss this?

I might have misscommunicated; I only meant that Soss's `advancedHMC()`doesn't let me pass an `AbstractRNG`. It looks like AdvancedHMC.jl does.

Yeah, the rng for `rand()` in Soss would make things easier too.

**Comment by DilumAluthge** [comment]
> > Could you open an issue on AdvancedHMC.jl to discuss this?
> 
> I might have miscommunicated; I only meant that Soss's `advancedHMC()`doesn't let me pass an `AbstractRNG`. It looks like AdvancedHMC.jl does.
> 
> Yeah, the rng for `rand()` in Soss would make things easier too.

Oh gotcha.

**Comment by cscherrer** [comment]
I'm working on adding rng argument in `rand.jl` FYI

File: test/readme.jl [hunk 1]
```diff
@@ -2,7 +2,8 @@ using Weave
 using Documenter
 
 function buildREADME()
-    weave("../README.jmd", doctype= "github", throw_errors=true, cache=:refresh, args=Dict(:seed => 6))
+    set_chunk_defaults!(:error=>false)
```
Comment (millerjoey) [inline test/readme.jl:5]: See
http://weavejl.mpastell.com/stable/chunk_options/#Evaluation

The `false` feels backwards to me, but nevertheless it now should stop if errors are thrown so the readme functions as a test.

**Comment by cscherrer** [comment]
Think I got it. Generated function stuff is confusing if you take any kind of break from it :)

```julia
julia> m = @model begin
           x ~ Normal()
       end;

julia> rand(m())
(x = -0.5424325612652338,)

julia> rand(StableRNG(1), m())
(x = -0.5325200748641231,)

julia> rand(StableRNG(1), m())
(x = -0.5325200748641231,)

julia> rand(StableRNG(2), m())
(x = -1.6747889454750067,)

julia> rand(StableRNG(2), m())
(x = -1.6747889454750067,)
```

**Comment by millerjoey** [comment]
Imagine how they look to the rest of us!

Do you think it's time we remove the Readme from the tests, or just copy it over to a separate test file (after fixing this error of course)?

**Comment by cscherrer** [comment]
> Do you think it's time we remove the Readme from the tests, or just copy it over to an explicit test file? It would help to speed things up, and might unnecessarily complicate this debug.

Are we down to just the README now? If that's the case then sure. I wanted to be sure the README was accurate, but at some point it should be high-level anyway, and details can go in the actual docs

**Comment by millerjoey** [comment]
The error was thrown during weaving the readme, and when I take it out the tests pass. Should something like this work now?

```julia
m = @model X begin
    β ~ Normal() |> iid(size(X,2))
    y ~ For(eachrow(X)) do x
        Normal(x' * β, 1)
    end
end

rng = StableRNG(42)
X = randn(rng, 6,2)
truth = rand(rng, m(X=X))
```

I'm in the process of moving some of the code in the readme to a test.

**Comment by millerjoey** [comment]
The test passes on my computer running 1.5, but fails on 1.4. I'm running the code on both now side by side.. it looks like `dynamicHMC` is producing slightly perturbed values of `β` samples. Code:
```julia
m = @model X begin
    β ~ Normal() |> iid(size(X,2))
    y ~ For(eachrow(X)) do x
        Normal(x' * β, 1)
    end
end

rng = StableRNG(42)
X = randn(rng, 6,2)
truth = rand(rng, m(X=X))
post = dynamicHMC(rng, m(X=X), (y=truth.y,))
pred = predictive(m,:β)

@testset "dynamicHMC" begin
    @test sum(v.β for v in post) == [137.9498847939604, 1253.8966977259338] # This test fails in Julia 1.4
    ...
end
```

`StableRNGs` seems to work right (`X` and `truth` are identical) but we get small differences of around 10e-5:

```julia
# Julia 1.5
julia> post = dynamicHMC(rng, m(X=X), (y=truth.y,))
1000-element Array{NamedTuple{(:β,),Tuple{Array{Float64,1}}},1}:
 (β = [0.17363181867461086, 0.9563576648513301],)
 (β = [-0.2943707196954747, 1.1813828508649298],)
 (β = [-0.25197558874192, 1.6232513960716974],)
 (β = [-0.05247573380112147, 1.1786089361807153],)
 (β = [0.28939536361928425, 1.2552721015608366],)
 (β = [-0.26386085165216705, 1.1416480060927239],)
```

```julia
# Julia 1.4
julia> post = dynamicHMC(rng, m(X=X), (y=truth.y,))
1000-element Array{NamedTuple{(:β,),Tuple{Array{Float64,1}}},1}:
 (β = [0.17365118251660278, 0.9563427680930621],)
 (β = [-0.2944620290466954, 1.1814124027620347],)
 (β = [-0.252037913198384, 1.6232974425744078],)
 (β = [-0.05248995773662024, 1.1786236748548977],)
 (β = [0.2893923877988596, 1.255287109143727],)
 (β = [-0.26388129199168536, 1.1416550697139776],)
```

It might relate to [DynamicHMC.jl issue 132](https://github.com/tpapp/DynamicHMC.jl/issues/132). Should we bump the tests to use Julia 1.5 and see if this problem doesn't come back, or dig into it?

**Comment by cscherrer** [comment]
Thanks @millerjoey ! I think focusing on 1.5 is fine for most of my use, but SossMLJ has more moving pieces so I'm not sure. @DilumAluthge what do you think?

**Comment by cscherrer** [comment]
I think this might be because, even if the stream of random numbers is identical, updates in algorithm implementations might change how those values are used, so we get different results.

If we want to go across several versions of Julia and packages, we may need to drop the nondeterministic doctests altogether, and just put this part in terms of approximate equality of some expected value

**Comment by millerjoey** [comment]
That makes sense. Maybe DynamicHMC.jl changed its algo a bit. `@test_nowarn` is also useful to verify that it doesn't error.

**Comment by cscherrer** [comment]
Tests pass!!! @millerjoey is this ready to go?

**Review by cscherrer — APPROVED** [review]

**Pull request merged**